common/io_lock.c

```diff
@@ -0,0 +1,75 @@
+#include "io_lock.h"
+#include <assert.h>
+#include <ccan/io/io_plan.h>
+
+struct io_lock {
+	bool locked;
+};
+
+/* Struct to hold information while we wait for the lock to be freed */
+struct io_lock_waiter {
+	struct io_plan *(*next)(struct io_conn *conn, void *next_arg);
+	void *arg;
+	struct io_lock *lock;
+	enum io_direction dir;
+};
+
+struct io_lock *io_lock_new(const tal_t *ctx)
+{
+	struct io_lock *lock = tal(ctx, struct io_lock);
+	lock->locked = false;
+	return lock;
+}
+
+static struct io_plan *io_lock_try_acquire(struct io_conn *conn,
+					   struct io_lock_waiter *waiter)
+{
+	/* Destructure waiter, since we might be freeing it below */
+	struct io_plan *(*next)(struct io_conn *, void *) = waiter->next;
+	void *next_arg = waiter->arg;
+
+	if (!waiter->lock->locked) {
+		waiter->lock->locked = true;
+		tal_free(waiter);
```

Comment:
I'm guessing because it's not necessarily freed in all branches?

Reply:
There are four kinds of *pointer) function parameters on this spectrum:
1. ones where ownership stays with the caller.  This is most common.
2. ones where the callee always takes ownership, eg `free(ptr)` obviously takes over ptr.
3. ones which sometimes take ownership, depending on internal conditions, like this one, or a reference-counting release function.  These are always dangerous (eg. realloc!), but sometimes necessary, esp. for internal functions handling internal objects.
4. ones which always take ownership if the caller tells them to, otherwise never take ownership.  This is what TAKES annotations mean; it takes the pointer IF, ONLY IF, and ALWAYS IF the caller hands it take(ptr).